### ISSUE TITLE
PyQt5 5.13.2 breaks macOS window textures/content if dark mode is enabled

### ISSUE THREAD

**Issue by crwood** [gridsync/gridsync#267]
PyQt 5.13.2 breaks window textures and content in Gridsync, causing windows to appear blank or transparent. The following error messages were observed (which may point to two separate issues):

```
2019-11-22 17:20:58.416 Gridsync[24524:2438424] *** WARNING: Textured window <QNSPanel: 0x7faf33598830; contentView=NSObject(0x0)> is getting an implicitly transparent titlebar. This will break when linking against newer SDKs. Use NSWindow's -titlebarAppearsTransparent=YES instead.
2019-11-22 17:20:58.471 Gridsync[24524:2438424] It does not make sense to draw an image when [NSGraphicsContext currentContext] is nil.  This is a programming error. Break on void _NSWarnForDrawingImageWithNoCurrentContext(void) to debug.  This will be logged only once.  This may break in the future.
```

I could not reproduce the issue on macOS Sierra (10.12) or High Sierra (10.13) but, on Mojave, downgrading/pinning PyQt5 to version 5.13.1 seemingly corrects the issue, resulting in window textures rendering correctly. As a result, Gridsync builds should temporarily(?) downgrade/pin the PyQt5 package until the code paths that cause the above two errors can be identified and fixed.

**Comment by crwood** [comment]
Re-opening to deploy a better fix later -- specifically, a fix that doesn't depend on keeping PyQt5 downgraded/pinned forever.

**Comment by MAKOMO** [comment]
I encountered a similar issue with my app, also build with py2app using PyQt 5.13.2 as reported here https://www.riverbankcomputing.com/pipermail/pyqt/2019-November/042305.html

For me removing the entries

	<key>NSRequiresAquaSystemAppearance</key>
	<string>false</string>

from the apps Info.plist resolved this, but deactivated the automatic dark mode support. Under 5.13.1 everything works as expected also with the lines above and dark mode is supported. Strange that this flag is not set in this apps  Info.plist.

**Comment by crwood** [comment]
Related PyInstaller issues:

https://github.com/pyinstaller/pyinstaller/issues/4615
https://github.com/pyinstaller/pyinstaller/issues/4627

**Comment by crwood** [comment]
Also (from @MAKOMO): https://www.riverbankcomputing.com/pipermail/pyqt/2019-December/042396.html

**Comment by crwood** [comment]
I can confirm that this behavior/issue is also present on macOS "Catalina" (10.15) with PyQt5 5.13.2, 5.14.0, and 5.14.1. Pinning/downgrading to 5.13.1 results in a working dark mode.

**Comment by crwood** [comment]
Possibly relevant:

According to the Qt5.13 [Supported Platforms page](https://doc.qt.io/archives/qt-5.13/supported-platforms.html):

>Support for specific configurations or operating system versions may end before the support for Qt 5.13 does. A subsequent patch release of Qt 5.13 may drop support for a configuration as a reference configuration or replace it with a currently-supported version.

This may account for why (Py)Qt5.13.1 works without issue while 5.13.2 is broken.

**Comment by crwood** [comment]
I spent some time investigating this and have figured out a solution. For anyone stumbling upon this issue in the future (including @MAKOMO!), the fix is twofold:

1. *Recompile both the PyInstaller bootloader and your Python interpreter(s) with a minimum of deployment target of "10.13"* (by setting `MACOSX_DEPLOYMENT_TARGET=10.13` and  appending `-mmacosx-version-min=10.13` to your `CFLAGS`/`CPPFLAGS`/`LDFLAGS`/`LINKFLAGS`). This ensures that your application will be backward-compatible with macOS 10.13 High Sierra -- which still receives security-updates from Apple. If you don't want or need to support 10.13, you can set this higher, but leaving it at 10.7 -- which is the [PyInstaller bootloader's default deployment-version](https://github.com/nehaljwani/pyinstaller/blob/5c09596322fed84c3cff0668c7d67c4a8852f1fb/bootloader/wscript#L548) -- causes the widgets to not render properly if `NSRequiresAquaSystemAppearance` is `False` (i.e., if dark mode support is enabled). [Here's how/where I recompile the PyInstaller bootloader in the Gridsync Makefile](https://github.com/gridsync/gridsync/blob/2752317f66bf67001ab19b75997354d1882c5b7f/Makefile#L167) and [here's how/where I compile my Python interpreters with pyenv when provisioning a builder via Vagrant/Virtualbox](https://github.com/gridsync/gridsync/blob/2752317f66bf67001ab19b75997354d1882c5b7f/vagrantfiles/macos-10.14/Vagrantfile#L43).

2. *Use an Xcode version of 11 or higher* when recompiling the above. This ensures that your deployed (PyInstaller/Python) binaries are actually compatible with Qt 5.14 (and 5.13.2!) and the dark mode improvements to that were introduced in that version. Prior to 5.13.2, Qt [accepted Xcode 9 and 10](https://doc.qt.io/archives/qt-5.13/supported-platforms.html) but, as of 5.14 (and 5.13.2), Qt [requires Xcode 11](https://doc.qt.io/qt-5/supported-platforms.html). Note that [Xcode 11 requires a Mac running macOS Mojave 10.14.4 or later](https://developer.apple.com/documentation/xcode_release_notes/xcode_11_release_notes). In my case, I built Gridsync on a macOS 10.14.6 with (Py)Qt 5.14 and an `NSRequiresAquaSystemAppearance` plist entry of `False` and was able to successfully launch the application on macOS 10.13, 10.14, and 10.15 with dark mode on or off.

PR forthcoming shortly...

**Comment by bjones1** [comment]
Thanks for investigating this -- nice work. What's the oldest, currently-supported version of OS X / MacOS out there? We want to compile the bootloader for this version, since most applications are forward-compatible, but not backward-compatible.

**Comment by crwood** [comment]
No problem! Glad the information can be useful!

The oldest, currently-supported version of macOS is 10.13 ("High Sierra") (which is [supported until September 2020](https://en.wikipedia.org/wiki/MacOS_High_Sierra)). Generally speaking, Apple has historically supported whatever the current OS version is plus the previous two -- so, currently, 10.15 ("Catalina") plus 10.14 ("Mojave") and 10.13 ("High Sierra"). Whenever a new macOS version is released -- typically in the fall -- the oldest version gets dropped in favor of the new one (so, at any given time, there are only three OS versions receiving security-updates).

**Comment by bjones1** [comment]
I can definitely see changing https://github.com/pyinstaller/pyinstaller/blob/5c09596322fed84c3cff0668c7d67c4a8852f1fb/bootloader/wscript#L548 to build on 10.13 today, but requiring Xcode 11 means dropping 10.13 support, from what I understand, which only makes sense after 10.13 EOL.

**Comment by crwood** [comment]
You don't actually need to drop 10.13 support; so long as you have `MACOSX_DEPLOYMENT_TARGET=10.13` (or higher) in your build environment, the resultant binary will work on 10.13 (or higher) -- or, at least, it's supposed to, according to Apple... You _will_, however, need a Mac that's running 10.14.4 or higher in order to run/use Xcode 11 though.

My suggestion: build your bootloader on macOS 10.14.6 -- I used [this Vagrant box](https://app.vagrantup.com/yzgyyang/boxes/macOS-10.14) when I tested it -- but make sure that you're setting `MACOSX_DEPLOYMENT_TARGET=10.13` -- that way you're able to support 10.13 and higher (i.e., all of the versions that Apple supports) while also being compatible with newer versions of Qt5.

**Comment by bjones1** [comment]
OK, thanks -- makes sense. A PyInstaller PR with updates to `.travis.yml` (to use XCode 11) and to `wscript` (to set the min version to 10.13) would be welcome.

**Comment by MAKOMO** [comment]
Thanks for documenting this. It seems that recompiling python is not necessary. I am running (using py2app) on the standard Python 3.8 build from python.org. The bootloader (apptemplate in py2app) needs to be recompiled. I also seems not to be necessary to set `NSRequiresAquaSystemAppearance` to `false` in Info.plist, at least not on my 10.14 and 10.15 systems (I do not have access to macOS 10.13 anymore).

**Comment by crwood** [comment]
Whether or not your python interpreter needs to be recompiled depends on the compatibility between a) the OS-version you're building on and b) the OS-version you're targeting (or running the application on). In my case, a `python3.7` interpreter that's compiled (and later frozen with PyInstaller) on a macOS 10.15 machine will fail to run on macOS 10.13 with the following traceback:

```
Traceback (most recent call last):
  File "PyInstaller/loader/rthooks/pyi_rth_pkgres.py", line 13, in <module>
  File "/Users/vagrant/buildbot/macOS_10_15/build/.tox/pyinstaller/lib/python3.7/site-packages/PyInstaller/loader/pyimod03_importers.py", line 489, in exec_module
  File "pkg_resources/__init__.py", line 33, in <module>
  File "/Users/vagrant/buildbot/macOS_10_15/build/.tox/pyinstaller/lib/python3.7/site-packages/PyInstaller/loader/pyimod03_importers.py", line 489, in exec_module
  File "platform.py", line 116, in <module>
  File "/Users/vagrant/buildbot/macOS_10_15/build/.tox/pyinstaller/lib/python3.7/site-packages/PyInstaller/loader/pyimod03_importers.py", line 489, in exec_module
  File "subprocess.py", line 153, in <module>
ImportError: dlopen(/Users/vagrant/Desktop/Gridsync.app/Contents/MacOS/select.cpython-37m-darwin.so, 2): Symbol not found: ____chkstk_darwin
  Referenced from: /Users/vagrant/Desktop/Gridsync.app/Contents/MacOS/select.cpython-37m-darwin.so (which was built for Mac OS X 10.15)
  Expected in: /usr/lib/libSystem.B.dylib
 in /Users/vagrant/Desktop/Gridsync.app/Contents/MacOS/select.cpython-37m-darwin.so
[1782] Failed to execute script pyi_rth_pkgres
```

Recompiling my python interpreter with `MACOSX_DEPLOYMENT_TARGET=10.13` causes this particular error to go away.

**Comment by jonathannaguin** [comment]
Any luck on creating that PR to PyInstaller? Are they any other changes needed in addition to the travis.yml and the wscript? Thanks!

**Comment by crwood** [comment]
For anyone still stumbling upon this issue, the PyInstaller/bootloader-related fix described above has now been [merged into the PyInstaller development branch](https://github.com/pyinstaller/pyinstaller/pull/5274). Thank you @BoboTiG!

Note, however, that you may still need to recompile your python interpreter(s) with `MACOSX_DEPLOYMENT_TARGET=10.13` (or higher) using Xcode 11 (or higher) in order to have fully working support for dark mode with PyQt5/PySide2 apps.

**Comment by BoboTiG** [comment]
I am not 100% sure that Python needs to be recompiled too. Maybe just PyInstaller should do the trick?
